examples_apps/run_minimizing_correlations.py

```diff
@@ -0,0 +1,355 @@
+#
+
+cutoff = 30
+
+
+######################################################################
+# Creating the initial states
+# ---------------------------
+#
+# We define our input state
+# :math:`\newcommand{ket}[1]{\left|#1\right\rangle}\ket{\psi}`, an equal
+# superposition of :math:`\ket{0}` and :math:`\ket{1}`:
+#
+# .. math:: \ket{\psi}=\frac{1}{\sqrt{2}}\left(\ket{0}+\ket{1}\right)
+#
+
+psi = np.zeros([cutoff], dtype=np.complex128)
+psi[0] = 1.0
+psi[1] = 1.0
+psi /= np.linalg.norm(psi)
+
+
+######################################################################
+# We can now define our initial random guess for the second state
+# :math:`\ket{\phi}`:
+#
+
+phi = np.random.random(size=[cutoff]) + 1j*np.random.random(size=[cutoff])
+phi[10:] = 0.
+phi /= np.linalg.norm(phi)
+
+
+######################################################################
+# Next, we define the creation operator :math:`\hat{a}`,
+#
+
+a = np.diag(np.sqrt(np.arange(1, cutoff)), k=1)
+
+
+######################################################################
+# the number operator :math:`\hat{n}=\hat{a}^\dagger \hat{a}`,
+#
+
+n_opt = a.T @ a
+
+
+######################################################################
+# and the quadrature operators :math:`\hat{x}=a+a^\dagger`,
+# :math:`\hat{p}=-i(a-a^\dagger)`.
+#
+
+# Define quadrature operators
+x = a + a.T
+p = -1j*(a-a.T)
+
+
+######################################################################
+# We can now calculate the displacement of the states in the phase space,
+# :math:`\alpha=\langle \psi \mid\hat{a}\mid\psi\rangle`. The following
+# function calculates this displacement, and then displaces the state by
+# :math:`-\alpha` to ensure it has zero displacement.
+#
+
+def recenter(state):
+    alpha = state.conj() @ a @ state
+    disp_alpha = expm(alpha.conj()*a - alpha*a.T)
+    out_state = disp_alpha @ state
+    return out_state
+
+
+######################################################################
+# First, let’s have a look at the displacement of state :math:`\ket{\psi}`
+# and state :math:`\ket{\phi}`:
+#
+
+print(psi.conj().T @ a @ psi)
+print(phi.conj().T @ a @ phi)
+
+
+######################################################################
+# Now, let’s center them in the phase space:
+#
+
+psi = recenter(psi)
+phi = recenter(phi)
+
+
+######################################################################
+# Checking they now have zero displacement:
+#
+
+print(np.round(psi.conj().T @ a @ psi, 9))
+print(np.round(phi.conj().T @ a @ phi, 9))
+
+
+######################################################################
+# Performing the optimization
+# ---------------------------
+#
+# We can construct the variational quantum circuit cost function, using Strawberry
+# Fields:
+#
+
+penalty_strength = 10
+
+prog = sf.Program(2)
+eng = sf.Engine("tf", backend_options={"cutoff_dim": cutoff})
+
+psi = tf.cast(psi, tf.complex64)
+phi_var_re = tf.Variable(phi.real)
+phi_var_im = tf.Variable(phi.imag)
```

Comment:
Are the imaginary and real parts separated just for clarity, or does are real variables somehow needed for the cost function below?

Reply:
Unfortunately, TF2.0 doesn't support optimizing complex variables, so it had to be separated. TF2.1 adds support again, but then breaks complex gradient computations 😆